Enable errcheck linter rule

vishwas-sharma2480 opened the pull request:
## Summary of Changes

- Enabled the `errcheck` linter rule.
- Reviewed and fixed all linter errors related to the `errcheck` rule.
- Updated the codebase to ensure that error return values are checked when necessary.
- Removed the line that disables the `errcheck` rule, enabling it for the project.

## Additional Information

- Closes: https://github.com/percona/pmm/issues/2346
- Enabled the `errcheck` rule to improve code quality and error handling.

**Note**: I raised this PR on a new branch since my last PR was made from a branch that was not updated with the main branch.

---

*This PR is based on a new branch created from the main branch to ensure it is up to date with the latest changes.*

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/pmm?pullRequest=2470) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/pmm?pullRequest=2470) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: vishwas-sharma2480<br/>:x: codesee-maps[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/pmm?pullRequest=2470) it.</sub>

artemgavrilov commented on .golangci.yml:
```diff
@@ -114,7 +114,7 @@ linters:
     - paralleltest
     - gocognit
-    - errcheck
+    #- errcheck
```
TODO:
```suggestion
```

vishwas-sharma2480 replied:
@artem I made the same changes as suggested

artemgavrilov replied:
Let's just remove this linter from this list instead of commenting it.

artemgavrilov commented on managed/services/dbaas/kubernetes/client/database/database.go:
```diff
@@ -50,9 +50,13 @@ func NewForConfig(c *rest.Config) (*DatabaseClusterClient, error) {
 
 	addToScheme.Do(func() {
-		dbaasv1.SchemeBuilder.AddToScheme(scheme.Scheme)
+		addToSchemeError=dbaasv1.SchemeBuilder.AddToScheme(scheme.Scheme)
 		metav1.AddToGroupVersion(scheme.Scheme, dbaasv1.GroupVersion)
 	})
+	if addToSchemeError != nil {
+		       return nil, addToSchemeError
+		   }
+
```
Let's just mute linter here. Whole `dbass` package will be dropped soon.

vishwas-sharma2480 replied:
Hi @artemgavrilov  i made the changes please review

vishwas-sharma2480 commented:
Hi Artem
Sure I Will make the changes and let you know

On Fri, 15 Sep 2023, 15:58 Artem Gavrilov, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In .golangci.yml
> <https://github.com/percona/pmm/pull/2470#discussion_r1327099252>:
>
> > @@ -114,7 +114,7 @@ linters:
>      - revive
>      - paralleltest
>      - gocognit
> -    - errcheck
> +    #- errcheck
>
> TODO:
> ⬇️ Suggested change
>
> -    #- errcheck
>
> ------------------------------
>
> In managed/services/dbaas/kubernetes/client/database/database.go
> <https://github.com/percona/pmm/pull/2470#discussion_r1327105979>:
>
> > -		dbaasv1.SchemeBuilder.AddToScheme(scheme.Scheme)
> +		addToSchemeError=dbaasv1.SchemeBuilder.AddToScheme(scheme.Scheme)
>  		metav1.AddToGroupVersion(scheme.Scheme, dbaasv1.GroupVersion)
>  	})
> +	if addToSchemeError != nil {
> +		       return nil, addToSchemeError
> +		   }
> +
>
> Let's just mute linter here. Whole dbass package will be dropped soon.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/percona/pmm/pull/2470#pullrequestreview-1628650947>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BAGMXGJIHSVEESCBJINWAX3X2QUTVANCNFSM6AAAAAA4SNCAAE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

artemgavrilov commented on managed/services/dbaas/kubernetes/client/database/database.go:
```diff
@@ -50,9 +50,11 @@ func NewForConfig(c *rest.Config) (*DatabaseClusterClient, error) {
+		dbaasv1.SchemeBuilder.AddToScheme(scheme.Scheme)//nolint:errcheck
 		metav1.AddToGroupVersion(scheme.Scheme, dbaasv1.GroupVersion)
 	})
+}
+
```
This bracket makes this file invalid
```suggestion
```

vishwas-sharma2480 replied:
Hi @artemgavrilov I made the changes

artemgavrilov reviewed (changes requested):
Tests and Checks are failing because of this extra bracket

vishwas-sharma2480 commented:
hi @artemgavrilov i tried the make format command in the  my local PMM directory but got this
make: bin/gofumpt: No such file or directory
make: *** [Makefile.include:155: format] Error 127

artemgavrilov commented:
> hi @artemgavrilov for go format i first did go init but needed to do go mod tidy to update the go.mod file but i am facing this issues like this a for a alot of modules example go: [github.com/percona/pmm/api-tests/server](https://github.com/percona/pmm/api-tests/server) tested by     [github.com/percona/pmm/api-tests/server.test](https://github.com/percona/pmm/api-tests/server.test) imports     [github.com/percona/pmm/api/platformpb/json/client/platform](https://github.com/percona/pmm/api/platformpb/json/client/platform): no matching versions for query "latest"

I assume that's because you have go v1.21 on your machine. Try this:
```
GOTOOLCHAIN=go1.20 make init
make format
```

artemgavrilov commented:
@vishwas-sharma2480 Nevermind, we just updated project go version up to 1.21. So just sync with the master and then try again `make init && make format`

vishwas-sharma2480 commented:
hi @artemgavrilov  now i am getting this issue
compile: version "go1.18.1" does not match go tool version "go1.21.0"
tools.go:47: running "go": exit status 1
make: *** [Makefile.include:20: init] Error 1

it seems now the compile version is 1.18 instead of 1.21

vishwas-sharma2480 commented:
@artemgavrilov  also I am encountering this issue as well
golangci-lint run
WARN [runner] The linter 'structcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
ERRO [linters_context] gocritic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt: setting an explicit GOROOT can fix this problem.

artemgavrilov commented:
> hi @artemgavrilov now i am getting this issue compile: version "go1.18.1" does not match go tool version "go1.21.0" tools.go:47: running "go": exit status 1 make: *** [Makefile.include:20: init] Error 1
> 
> it seems now the compile version is 1.18 instead of 1.21

Nope, project definitely uses 1.21, you can check it in go.mod file in `main` branch.

Check go version on your machine `go version`.

vishwas-sharma2480 commented:
compile: version "go1.18.1" does not match go tool version "go1.20"
tools.go:47: running "go": exit status 1
make: *** [Makefile.include:20: init] Error 1
compile: version "go1.18.1" does not match go tool version "go1.19"
tools.go:47: running "go": exit status 2
make: *** [Makefile.include:20: init] Error 1

i tried and use different versions got the same output
and ths was when i used 1.18.1
note: module requires Go 1.19
tools.go:48: running "go": exit status 2
make: *** [Makefile.include:20: init] Error 1

also this is my current version but got the same issue
vboxuser@ubuntu1:~/pmm$ go version
go version go1.21.0 linux/amd64
tools.go:47: running "go": exit status 1
make: *** [Makefile.include:20: init] Error 1

artemgavrilov commented:
@vishwas-sharma2480 You are doing in in devcontainer or right on the host machine? I mean do you use `make env-up, make env` commands? If not then something wrong with your environment, go1.18.1 is pretty old, we migrated from it while time ago and there is no even mention of it in project sources.

If you are using devcontainer then try to update it with `make env-update-image`

vishwas-sharma2480 commented:
i am working locally on host machine

artemgavrilov commented:
> i am working locally on host machine

Check your `GOROOT` env variable? Is it pointing to the proper version?

https://stackoverflow.com/questions/46693653/compile-version-go1-9-does-not-match-go-tool-version-go1-9-1

vishwas-sharma2480 commented:
I think I fixedit but now facing this issue
make init
rm -rf bin/*
cd tools && go generate -x -tags=tools
go build -o ../bin/benchstat [golang.org/x/perf/cmd/benchstat](http://golang.org/x/perf/cmd/benchstat)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/main.go:398:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/main.go:398:2): package flag is not in std (/usr/bin/src/flag)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/main.go:399:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/main.go:399:2): package fmt is not in std (/usr/bin/src/fmt)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:8:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:8:2): package bufio is not in std (/usr/bin/src/bufio)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:9:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:9:2): package bytes is not in std (/usr/bin/src/bytes)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/internal/bytesconv/atoi.go:8:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/internal/bytesconv/atoi.go:8:2): package errors is not in std (/usr/bin/src/errors)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/internal/bytesconv/atof.go:13:8](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/internal/bytesconv/atof.go:13:8): package math is not in std (/usr/bin/src/math)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/internal/bytesconv/atoi.go:10:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/internal/bytesconv/atoi.go:10:2): package strconv is not in std (/usr/bin/src/strconv)
../../go/pkg/mod/[github.com/aclements/go-moremath@v0.0.0-20210112150236-f10218a38794/stats/dist.go:7:8](http://github.com/aclements/go-moremath@v0.0.0-20210112150236-f10218a38794/stats/dist.go:7:8): package math/rand is not in std (/usr/bin/src/math/rand)
../../go/pkg/mod/[github.com/aclements/go-moremath@v0.0.0-20210112150236-f10218a38794/stats/sample.go:9:2](http://github.com/aclements/go-moremath@v0.0.0-20210112150236-f10218a38794/stats/sample.go:9:2): package sort is not in std (/usr/bin/src/sort)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchmath/anone.go:10:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchmath/anone.go:10:2): package sync is not in std (/usr/bin/src/sync)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchunit/tidy.go:8:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchunit/tidy.go:8:2): package strings is not in std (/usr/bin/src/strings)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchunit/parse.go:11:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchunit/parse.go:11:2): package unicode is not in std (/usr/bin/src/unicode)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:11:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:11:2): package io is not in std (/usr/bin/src/io)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/files.go:9:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/files.go:9:2): package os is not in std (/usr/bin/src/os)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:14:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchfmt/reader.go:14:2): package unicode/utf8 is not in std (/usr/bin/src/unicode/utf8)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchproc/internal/parse/tok.go:10:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchproc/internal/parse/tok.go:10:2): package regexp is not in std (/usr/bin/src/regexp)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchproc/projection.go:9:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/benchproc/projection.go:9:2): package hash/maphash is not in std (/usr/bin/src/hash/maphash)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/internal/benchtab/builder.go:9:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/internal/benchtab/builder.go:9:2): package encoding/csv is not in std (/usr/bin/src/encoding/csv)
../../go/pkg/mod/[golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/internal/benchtab/builder.go:14:2](http://golang.org/x/perf@v0.0.0-20230717203022-1ba3a21238c9/cmd/benchstat/internal/benchtab/builder.go:14:2): package runtime is not in std (/usr/bin/src/runtime)
tools.go:47: running "go": exit status 1
make: *** [Makefile.include:20: init] Error 1

artemgavrilov commented:
@vishwas-sharma2480 It looks like your GOPATH now pointing to the wrong golang std lib version. Try to reinstall golang completey on your machine (first remove it completely).

vishwas-sharma2480 commented:
hi@ artem i  did what you suggeted I got this output from make init
make init
rm -rf bin/*
cd tools && go generate -x -tags=tools
go build -o ../bin/benchstat [golang.org/x/perf/cmd/benchstat](http://golang.org/x/perf/cmd/benchstat)
go build -o ../bin/buf [github.com/bufbuild/buf/cmd/buf](http://github.com/bufbuild/buf/cmd/buf)
go build -o ../bin/dlv [github.com/go-delve/delve/cmd/dlv](http://github.com/go-delve/delve/cmd/dlv)
go build -o ../bin/gci [github.com/daixiang0/gci](http://github.com/daixiang0/gci)
go build -o ../bin/go-consistent [github.com/quasilyte/go-consistent](http://github.com/quasilyte/go-consistent)
go build -o ../bin/go-junit-report [github.com/jstemmer/go-junit-report](http://github.com/jstemmer/go-junit-report)
go build -o ../bin/go-sumtype [github.com/BurntSushi/go-sumtype](http://github.com/BurntSushi/go-sumtype)
go build -o ../bin/gofumpt [mvdan.cc/gofumpt](http://mvdan.cc/gofumpt)
go build -o ../bin/goimports [golang.org/x/tools/cmd/goimports](http://golang.org/x/tools/cmd/goimports)
go build -o ../bin/mockery [github.com/vektra/mockery/v2](http://github.com/vektra/mockery/v2)
go build -o ../bin/protoc-gen-go [google.golang.org/protobuf/cmd/protoc-gen-go](http://google.golang.org/protobuf/cmd/protoc-gen-go)
go build -o ../bin/protoc-gen-go-grpc [google.golang.org/grpc/cmd/protoc-gen-go-grpc](http://google.golang.org/grpc/cmd/protoc-gen-go-grpc)
go build -o ../bin/protoc-gen-grpc-gateway [github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-grpc-gateway](http://github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-grpc-gateway)
go build -o ../bin/protoc-gen-openapiv2 [github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-openapiv2](http://github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-openapiv2)
go build -o ../bin/reform [gopkg.in/reform.v1/reform](http://gopkg.in/reform.v1/reform)
go build -o ../bin/reviewdog [github.com/reviewdog/reviewdog/cmd/reviewdog](http://github.com/reviewdog/reviewdog/cmd/reviewdog)
go build -o ../bin/swagger [github.com/go-swagger/go-swagger/cmd/swagger](http://github.com/go-swagger/go-swagger/cmd/swagger)
go build -o ../bin/swagger-order [github.com/Percona-Lab/swagger-order](http://github.com/Percona-Lab/swagger-order)
go build -o ../bin/protoc-gen-validate [github.com/envoyproxy/protoc-gen-validate](http://github.com/envoyproxy/protoc-gen-validate)
go build -o ../bin/ifacemaker [github.com/vburenin/ifacemaker](http://github.com/vburenin/ifacemaker)
env CGO_ENABLED=0 go build -o ../bin/license-eye [github.com/apache/skywalking-eyes/cmd/license-eye](http://github.com/apache/skywalking-eyes/cmd/license-eye)
# Install golangci-lint
curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b ./bin v1.51.2 # Version should match specified in CI
curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to [raw.githubusercontent.com:443](http://raw.githubusercontent.com:443/)

vishwas-sharma2480 commented:
I was able t9 resolve it commited the changes made by make format

vishwas-sharma2480 commented:
Hi @artemgavrilov 
can please review the PR and let me know if any further changes needed

vishwas-sharma2480 commented:
Hi @artemgavrilov can you please review the PR and let me know if any changes reqired

ademidoff commented:
@vishwas-sharma2480 From what I can see in the logs, there are more `errcheck` warnings to be fixed. Some examples:

- https://github.com/percona/pmm/pull/2470/files#file-agent-commands-run-go-L112
- https://github.com/percona/pmm/pull/2470/files#file-managed-services-dbaas-kubernetes-client-client-go-L233
- https://github.com/percona/pmm/pull/2470/files#file-managed-services-dbaas-kubernetes-client-client-go-L570
- https://github.com/percona/pmm/pull/2470/files#file-managed-services-management-dbaas-components_service-go-L540
- https://github.com/percona/pmm/pull/2470/files#file-managed-services-management-dbaas-version_service_client-go-L161

vishwas-sharma2480 commented:
Hi @ademidoff i esolved the requested changes can please review and let me know

ademidoff commented:
> Hi @ademidoff i esolved the requested changes can please review and let me know

Thanks! Now the checks don't pass :| 
You can make them by either running `make format` or by manually adding a space, like in https://github.com/percona/pmm/actions/runs/6462583867/job/17544435518?pr=2470#step:10:107

vishwas-sharma2480 commented:
hi @ademidoff I made the changes can you please review and let me know
